src/data/polynomial/reverse.lean

```diff
@@ -0,0 +1,214 @@
+
+@[simp] lemma rev_at_invol {N n : ℕ} : rev_at N (rev_at N n) = n :=
+begin
+  unfold rev_at,
+  split_ifs with h j,
+  { exact nat.sub_sub_self h, },
+  { exfalso,
+    apply j,
+    exact nat.sub_le N n, },
+  { refl, },
+end
+
+@[simp] lemma rev_at_small {N n : ℕ} (H : n ≤ N) : rev_at N n = N-n :=
+begin
+  unfold rev_at,
+  split_ifs,
+  refl,
+end
+
+/-- The function `reflect` of a natural number `N` and a polynomial `f`, applies the function
+`rev_at` to the exponents of the terms appearing in the expansion of `f`.  In practice, `reflect`
+is only used when `N` is at least as large as the degree of `f`.  Eventually, it will be used with
+`N` exactly equal to the degree of `f`.  -/
+def reflect : ℕ → polynomial R → polynomial R :=
+ λ N : ℕ, λ f : polynomial R, ⟨ image (rev_at N)  (f.support), λ i : ℕ, f.coeff (rev_at N i),
+ begin
+  intro,
+  rw mem_image,
+  split,
+  { intro h,
+    rcases h with ⟨ a, ha, rfl⟩,
+    rwa [rev_at_invol, ← mem_support_iff_coeff_ne_zero], },
+  { intro,
+    use (rev_at N a),
+    rwa [rev_at_invol, eq_self_iff_true, and_true, mem_support_iff_coeff_ne_zero], },
+end ⟩
+
+@[simp] lemma reflect_zero {n : ℕ} : reflect n (0 : polynomial R) = 0 := rfl
+
+@[simp] lemma reflect_zero_iff {N : ℕ} (f : polynomial R) :
+ reflect N (f : polynomial R) = 0 ↔ f=0 :=
+begin
+  split,
+  { intros a,
+    injection a with f0 f1,
+    rwa [image_eq_empty, support_eq_empty] at f0, },
+  { intro a,
+    rw [a, reflect_zero], },
+end
+
+@[simp] lemma reflect_add (f g : polynomial R) (n : ℕ) :
+ reflect n (f+g) = reflect n f + reflect n g :=
+begin
+  ext1,
+  refl,
+end
+
+@[simp] lemma reflect_smul (f : polynomial R) (r : R) (N : ℕ) :
+ reflect N (C r * f) = C r * (reflect N f) :=
+begin
+  ext1,
+  unfold reflect,
+  rw [coeff_mk, coeff_C_mul, coeff_C_mul, coeff_mk],
+end
+
+@[simp] lemma reflect_C_mul_X_pow (N n : ℕ) {c : R} :
+ reflect N (C c * X ^ n) = C c * X ^ (rev_at N n) :=
+begin
+  ext1 m,
+  unfold reflect,
+  rw coeff_mk,
+  by_cases h : rev_at N n = m,
+  { rw [h, coeff_C_mul, coeff_C_mul, coeff_X_pow_self, ← h, rev_at_invol, coeff_X_pow_self], },
+  { rw not_mem_support_iff_coeff_zero.mp,
+    { symmetry,
+      apply not_mem_support_iff_coeff_zero.mp,
+      intro,
+      apply h,
+      exact (mem_support_C_mul_X_pow a).symm, },
+    { intro,
+      apply h,
+      rw ← @rev_at_invol N m,
+      apply congr_arg _,
+      exact (mem_support_C_mul_X_pow a).symm, }, },
+end
+
+@[simp] lemma reflect_monomial (N n : ℕ) : reflect N ((X : polynomial R) ^ n) = X ^ (rev_at N n) :=
+by rw [← one_mul (X^n), ← one_mul (X^(rev_at N n)), ← C_1, reflect_C_mul_X_pow]
+
+/-- The reverse of a polynomial f is the polynomial obtained by "reading f backwards".
+Even though this is not the actual definition, reverse f = f (1/X) * X ^ f.nat_degree. -/
+def reverse : polynomial R → polynomial R := λ f, reflect f.nat_degree f
+
+lemma pol_ind_Rhom_prod_on_card (cf cg : ℕ) {rp : ℕ → polynomial R → polynomial R}
```

Comment:
Sorry, I don't understand the name and purpose of this lemma. It seems like `rp_add`, `rp_smul` and `rp_mon` uniquely characterise `reflect`, and we have already proven them by this point, so why not put `reflect` in the statement and inline the proofs?

Reply:
I think that you are right.  I arrived at this lemma after a long sequence of steps and once I could prove one of the versions, I simply went with what it was.  I had not realized that I had proved a "more general lemma" that was actually equivalent to what I wanted to prove!

Reply:
I will reformulate it, so that this is already the main statement.